specs/agents/metrics-health.md

```diff
@@ -0,0 +1,129 @@
+
+### Event Queue Metrics
+
+The following metrics SHOULD be recorded by agents regarding the state of the internal event reporting queue:
+
+ - `agent.events.queue.min_size.pct`: The smallest queue usage in percent (range [0-1]) observed since this metric was reported the last time
+ - `agent.events.queue.max_size.pct`: The biggest queue usage in percent (range [0-1]) observed since this metric was reported the last time
+
+We capture both `min_size` and `max_size` to allow to distinguish between reasons for why the queue is becoming full:
+ * There is too much data captured / network errors: Both `min_size` and `max_size` are at `100%`
+ * The data is collected in bursts for which the queue is not correctly sized: `max_size` is at `100%`, but `min_size` is low.
+
+The queue usage can be computed based on how the agent defines the queue capacity.
+E.g. if the queue capacity is a fixed number of events, the usage can be computed based on the current number of events.
+If the queue capacity is in bytes, the usage can be computed based on the number of bytes currently occupied in the queue.
+
+All event queue metrics can be disabled via the `disable_metrics` configuration.
+
+#### Possible implementation for agents
+
+In order to track these metrics, most likely a custom implementation is required on top of the underlying queue.
+
+One approach can be to have two variables `min_size` and `max_size`, which are atomically updated on changes to the queue:
+ * When an item is added to the queue, update `max_size` with the current queue size if it is greater than `max_size`
+ * When an item is removed from the queue, update `min_size` with the current queue size if it is less than `min_size`
+ * When an item cannot be added to the queue (dropped), set `max_size` to `queue_capacity`
+
+After the metrics are exported, `min_size` and `max_size` need to be reset tu the current queue size.
+
+An atomic min/max can be implemented using the following algorithm:
+```java
+AtomicLong maxSize = new AtomicLong();
+
+public void updateMaxSize(long currentSize) {
+    //alternatively abort after like 10 iterations for constant latency
+    while (true) { 
+        long current = maxSize.get();
+        if (current >= currentSize) {
+            return;
+        }
+        boolean casSuccess = maxSize.compareAndSet(current, currentSize);
+        if (casSuccess) {
+            return;
+        }
+    }
+}
+
+```
+
+### Event Request Metrics
+
+Agents SHOULD expose the following metrics regarding Intake API networking:
+
+ - `agent.events.requests.count`: `COUNTER` of the number of requests made or attempted to the Intake API of the APM server
+ - `agent.events.requests.bytes`: `COUNTER` of the approximate number of bytes on the wire sent to the Intake API of the APM
+
+All these metrics MUST have the label `success`, which can have the values `true` or `false`. A request is counted with `success=true` iff the APM Server responded with `2xx`.
```

Comment:
```suggestion
The `agent.events.requests.*` metrics MUST have the label `success`, which can have the values `true` or `false`. A request is counted with `success=true` if the APM Server responded with `2xx`.
```

Reply:
`iff` stands for `if and only if` here, therefore it is not a typo.